### DISCUSSION TITLE
Is it a good idea to use egui to create a text editor

### DISCUSSION THREAD

**Discussion by rawhuul** [emilk/egui#3350] [Q&A] [answered]
Is it a good idea to use Egui or immediate mode GUI framework in general to create a simple minimal text editor which would provide basic syntax highlighting using package like tree-sitter and completion using LSP. I’m asking this because exchanging messages with LSP servers would be asynchronous. So is it a good idea?

**Comment by emilk** [comment, marked as answer]
It should work, yes, but be mindful that egui is currently not well suited for large text documents (1MB+); see for instance https://github.com/emilk/egui/issues/3086

**Reply by rawhuul** [reply]
Thanks for pointing it out, i’ll look into it

**Comment by rawhuul** [comment]
Hey there, I am using `egui` to make a very simple text editor and using `Rope` as backend data structure from `crop` crate. To use it, I have to implement `TextBuffer` trait which needs [`as_str`](https://docs.rs/egui/latest/src/egui/widgets/text_edit/text_buffer.rs.html#12) method to be implemented. I want to know in what conditions these method is invoked and for what, as I am using [`Rope`](https://docs.rs/crop/0.4.1/crop/struct.Rope.html) which doesn't implement `as_str`, so I have to do it in hacky way:

```rust
struct RopeText(Rope);

impl TextBuffer for RopeText {
    ...

    fn as_str(&self) -> &str {
        let string = self.0.to_string();
        let boxed_str = string.into_boxed_str();
        Box::leak(boxed_str)
    }

    ...
}
```

It is not very idle solution as I am converting the entire `Rope` to be a `String` (doesn't it undermines the whole idea of using `Rope`?), then leaking it. Is `as_str` invoked to render the text to widget? If that's the case, shouldn't the trait have method called `display` which could be like:

```rust
trait TextBuffer{
    fn display(&self) -> impl std::fmt::Display;
}
```
